lxd/api_project.go

```diff
@@ -195,56 +195,27 @@ func projectsGet(d *Daemon, r *http.Request) response.Response {
 // projectUsedBy returns a list of URLs for all instances, images, profiles,
 // storage volumes, networks, and acls that use this project.
 func projectUsedBy(ctx context.Context, tx *db.ClusterTx, project *cluster.Project) ([]string, error) {
-	usedBy := []string{}
-	instances, err := cluster.GetInstances(ctx, tx.Tx(), cluster.InstanceFilter{Project: &project.Name})
-	if err != nil {
-		return nil, err
-	}
-
-	profiles, err := cluster.GetProfiles(ctx, tx.Tx(), cluster.ProfileFilter{Project: &project.Name})
-	if err != nil {
-		return nil, err
-	}
-
-	images, err := cluster.GetImages(ctx, tx.Tx(), cluster.ImageFilter{Project: &project.Name})
-	if err != nil {
-		return nil, err
-	}
-
-	for _, instance := range instances {
-		apiInstance := api.Instance{Name: instance.Name}
-		usedBy = append(usedBy, apiInstance.URL(version.APIVersion, project.Name).String())
-	}
-
-	for _, profile := range profiles {
-		apiProfile := api.Profile{Name: profile.Name}
-		usedBy = append(usedBy, apiProfile.URL(version.APIVersion, project.Name).String())
-	}
-
-	for _, image := range images {
-		apiImage := api.Image{Fingerprint: image.Fingerprint}
-		usedBy = append(usedBy, apiImage.URL(version.APIVersion, project.Name).String())
+	reportedEntityTypes := []entity.Type{
+		entity.TypeInstance,
+		entity.TypeProfile,
+		entity.TypeImage,
+		entity.TypeStorageVolume,
+		entity.TypeNetwork,
+		entity.TypeNetworkACL,
 	}
 
-	volumes, err := tx.GetStorageVolumeURIs(ctx, project.Name)
+	entityURLs, err := cluster.GetEntityURLs(ctx, tx.Tx(), project.Name, reportedEntityTypes...)
```

Comment:
Could we use this pattern for other used by listings perhaps?

Reply:
Most likely yes, but we'll need to write new queries for each entity type that has a "used-by" and each entity type that is displayed in those lists.